[ZEPPELIN-704] Display elapse time for long running paragraph

sagarkulkarni3592 opened the pull request:
### What is this PR for?
For long running paragraph, it is hard to know how long it has already been running. We should either have the elapsed time displayed


### What type of PR is it?
[Bug Fix ]

### Todos


### What is the Jira issue?
ZEPPELIN-704 (https://issues.apache.org/jira/browse/ZEPPELIN-704)

### How should this be tested?
-Open a note.
-Run a paragraph , It should show you the elapsed time.


### Screenshots (if appropriate)
![screen shot 2016-04-28 at 3 46 22 pm](https://cloud.githubusercontent.com/assets/12127192/14882841/71313350-0d58-11e6-9fb5-0704e53f7746.png)

### Questions:
* Does the licenses files need update?No
* Is there breaking changes for older versions?No
* Does this needs documentation?No

Leemoonsoo commented:
Thanks @sagarkulkarni3592 for very useful feature.

It works nicely with 'default' looknfeel. However, 'simple' looknfeel it displays elapsed time above the progress bar on left side. like

![image](https://cloud.githubusercontent.com/assets/1540981/14933784/c8049f2c-0e4e-11e6-9283-b26fd5f56c90.png)

Could this be displayed below the progress bar on the right side in 'simple' looknfeel? the same position that "Took XX seconds" message placed.

![image](https://cloud.githubusercontent.com/assets/1540981/14933799/24fb0fa4-0e4f-11e6-9c71-10134dde7969.png)

sagarkulkarni3592 commented:
@Leemoonsoo Thank you for your suggestion.
I have done the necessary changes to make it appear on the same position that "Took XX seconds" message placed.
![screen shot 2016-05-03 at 12 52 02 pm](https://cloud.githubusercontent.com/assets/12127192/14976986/f89b8330-112d-11e6-8e58-d4a2a371d92a.png)

Leemoonsoo commented:
@sagarkulkarni3592 Thanks for taking care.
I tried the last commit and i can see text is displayed on the right corner in 'simple' looknfeel.
However, "Took xx seconds" and "xx seconds elapsed" are still displayed in the different place. Could you make it display at exactly the same place where "Took xx seconds" positioned?

![elaspsed](https://cloud.githubusercontent.com/assets/1540981/15025569/ba52c928-11ee-11e6-8527-673b66e4175c.gif)

sagarkulkarni3592 commented:
@Leemoonsoo Thanks for the suggestion.
Sorry for the delay. I have done the necessary changes.

![may-10-2016 10-46-39](https://cloud.githubusercontent.com/assets/12127192/15136215/d172feb8-169c-11e6-85cf-187ae3d5379a.gif)
![may-10-2016 10-47-49](https://cloud.githubusercontent.com/assets/12127192/15136216/d173ee04-169c-11e6-94eb-b69372225f54.gif)

Had to change property of Simple.css top: -9px from -13px as it was appearing on the loading blue line.

Is this good?

prabhjyotsingh commented:
Text floats outside paragraph

<img width="1090" alt="screen shot 2016-05-10 at 12 38 04 pm" src="https://cloud.githubusercontent.com/assets/674497/15138233/1a1b8978-16ac-11e6-9a6d-6b8d8d6d2f1b.png">

sagarkulkarni3592 commented:
Even after merging the pull request for [ZEPPELIN](https://github.com/apache/incubator-zeppelin/pull/848) , it was showing "Took xx seconds. Last updated by undefined at time XXX".
Fixed this to "Took xx seconds. Last updated by anonymous at time XXX".

sagarkulkarni3592 commented:
@prabhjyotsingh Thank u for the suggestion about text float.
I have made the necessary changes. Is this ok?
![screen shot 2016-05-10 at 3 27 14 pm](https://cloud.githubusercontent.com/assets/12127192/15142947/eceb4d3a-16c4-11e6-9b16-69017e453978.png)

prabhjyotsingh commented:
If there is difference in timezone for client and zeppelin server then UI shows a wrong value.

![virtualbox_windows 7_13_05_2016_16_09_06](https://cloud.githubusercontent.com/assets/674497/15245591/2844d290-1925-11e6-9be3-2e191f7dee00.png)

sagarkulkarni3592 commented:
@prabhjyotsingh Thank you for the bug. Fixed this using standard ISO time format. Tested with different time zones.
![screen shot 2016-05-23 at 5 38 50 pm](https://cloud.githubusercontent.com/assets/12127192/15470274/55c6cb16-210d-11e6-9dd7-e17a43fd6eef.png)
Checkout the time zones.
![screen shot 2016-05-23 at 5 42 02 pm](https://cloud.githubusercontent.com/assets/12127192/15470344/c466e3e4-210d-11e6-8ec5-819a5139c410.png)

This also changes the time format (customizable in code) at statement "last updated by xx at <time>".
![screen shot 2016-05-23 at 5 16 25 pm](https://cloud.githubusercontent.com/assets/12127192/15470100/f69cd2e4-210b-11e6-9a47-403ac54b85d7.png)
 
Added moment.js : Now, took xx seconds text will show as per moment.js [documentattion](http://momentjs.com/docs/#/displaying/fromnow/). Same is the case for elapsed time.
![screen shot 2016-05-23 at 5 25 34 pm](https://cloud.githubusercontent.com/assets/12127192/15470129/2594acfc-210c-11e6-87ee-90b3bb98e474.png)
![screen shot 2016-05-23 at 5 25 49 pm](https://cloud.githubusercontent.com/assets/12127192/15470127/25939060-210c-11e6-846e-3ce3ac4b9839.png)
![screen shot 2016-05-23 at 5 30 33 pm](https://cloud.githubusercontent.com/assets/12127192/15470128/2593d2a0-210c-11e6-9a34-5d277fa6934f.png)

prabhjyotsingh commented on zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js:
```diff
@@ -971,18 +971,23 @@ angular.module('zeppelinWebApp')
-    var desc = 'Took ' + (timeMs/1000) + ' seconds. Last updated by ' + user + ' at time ' + dateUpdated + '.';
+    var desc = 'Took '
+    + moment.duration(moment(pdata.dateFinished).diff(moment(pdata.dateStarted))).humanize()
+    + '. Last updated by ' + user + ' at ' + $rootScope.dateToString(dateUpdated) + '.';
```
Grunt build fails

    [INFO] src/app/notebook/paragraph/paragraph.controller.js
    [INFO]   line 979  col 5  Bad line breaking before '+'.
    [INFO]   line 980  col 5  Bad line breaking before '+'.

prabhjyotsingh commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ZeppelinIT.java:
```diff
@@ -200,13 +200,13 @@ public void testSparkInterpreterDependencyLoading() throws Exception {
 
       // add new dependency to spark interpreter
-      driver.findElement(By.xpath("//div[h3[text()[contains(.,'spark')]]]//button[contains(.,'edit')]")).sendKeys(Keys.ENTER);
+      driver.findElement(By.xpath("//div[h3[text()[contains(.,'%spark')]]]//button[contains(.,'edit')]")).sendKeys(Keys.ENTER);
```
This is fine without %spark, revert this.

prabhjyotsingh commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ZeppelinIT.java:
```diff
@@ -236,11 +236,11 @@ public void testSparkInterpreterDependencyLoading() throws Exception {
       // reset dependency
       interpreterLink.click();
-      driver.findElement(By.xpath("//div[h3[text()[contains(.,'spark')]]]//button[contains(.,'edit')]")).sendKeys(Keys.ENTER);
+      driver.findElement(By.xpath("//div[h3[text()[contains(.,'%spark')]]]//button[contains(.,'edit')]")).sendKeys(Keys.ENTER);
```
Same thing here, this is fine. Revert this.

corneadoug commented on zeppelin-web/src/components/navbar/navbar.controller.js:
```diff
@@ -115,4 +115,8 @@ angular.module('zeppelinWebApp').controller('NavCtrl', function($scope, $rootSco
   vm.loadNotes();
   $scope.checkUsername();
 
+  $rootScope.dateToString = function(date) {
```
This is not the right place for this function, also since its used only once in the code base, you could just use it inline.

sagarkulkarni3592 replied:
I thought we can have a common function for date to string which whole application can use as a standard date format.

corneadoug replied:
In that case it would be through a service, not the rootScope.
(The elapsed time function is a bit of a special case)

Overall since moment.js library call is already quite direct, there isn't really a need for it.

corneadoug commented:
I would prefer keeping the `elapsed` wording, otherwise it doesn't give indications that it is for the running process.

sagarkulkarni3592 commented:
@corneadoug Thank you for the suggestion.
For a long running process, say running for more than hour, it will show something like "3601 seconds elapsed". The number in seconds does not look good. What will be your suggestion on this?

corneadoug commented:
@sagarkulkarni3592 if you follow moment.js [documentation](http://momentjs.com/docs/#/displaying/fromnow/)
You can use `fromNow(true)` to remove the `ago`

sagarkulkarni3592 commented:
@corneadoug So, it will be "a few seconds elapsed" or "2 minutes elapsed". It this good?

corneadoug commented:
@sagarkulkarni3592 "Started 2 minutes ago" looks more gracious :)

sagarkulkarni3592 commented:
@corneadoug Latest commit have both the changes you suggested. :)
![screen shot 2016-05-25 at 12 03 12 pm](https://cloud.githubusercontent.com/assets/12127192/15530257/f6b9d2ac-2270-11e6-80d9-769a5e44c3a2.png)
@prabhjyotsingh corrected the test cases. :)

prabhjyotsingh commented:
Tested, works well. LGTM!

corneadoug commented:
Tested, Merging if there is no more discussions

bzz commented:
Merging as it includes HOTFIX for failing Integration Tests

bzz commented:
@sagarkulkarni3592 could you please assign a JIRA issue to yourself?
